share/pvss/pvss.go

```diff
@@ -31,64 +31,78 @@ type Suite interface {
+
+// PubVSSData holds all the public data produced by the dealer
+type PubVSSData struct {
+	N       int            // Number of participants
+	T       int            // Threshold to decrypt the secret
+	X       []kyber.Point  // Participants public keys
+	Shares  []*PubVerShare // Participants shares
+	PubPoly *share.PubPoly // Commited polynomial
+}
 
 // PubVerShare is a public verifiable share.
 type PubVerShare struct {
-	S share.PubShare // Share
-	P dleq.Proof     // Proof
+	S   share.PubShare // Share
+	Com kyber.Point    // Commitment of p(i)
+	P   dleq.Proof     // Proof
 }
 
 // EncShares creates a list of encrypted publicly verifiable PVSS shares for
 // the given secret and the list of public keys X using the sharing threshold
 // t and the base point H. The function returns the list of shares and the
 // public commitment polynomial.
-func EncShares(suite Suite, H kyber.Point, X []kyber.Point, secret kyber.Scalar, t int) (shares []*PubVerShare, commit *share.PubPoly, err error) {
-	n := len(X)
-	encShares := make([]*PubVerShare, n)
+func EncShares(suite Suite, H kyber.Point, X []kyber.Point, secret kyber.Scalar, t int) (*PubVSSData, error) {
+	p := new(PubVSSData)
+	p.N = len(X)
+	p.T = t
+	p.Shares = make([]*PubVerShare, p.N)
+	p.X = X
 
 	// Create secret sharing polynomial
 	priPoly := share.NewPriPoly(suite, t, secret, suite.RandomStream())
 
 	// Create secret set of shares
-	priShares := priPoly.Shares(n)
+	priShares := priPoly.Shares(p.N)
 
 	// Create public polynomial commitments with respect to basis H
-	pubPoly := priPoly.Commit(H)
+	p.PubPoly = priPoly.Commit(H)
 
 	// Prepare data for encryption consistency proofs ...
-	indices := make([]int, n)
-	values := make([]kyber.Scalar, n)
-	HS := make([]kyber.Point, n)
-	for i := 0; i < n; i++ {
+	indices := make([]int, p.N)
+	values := make([]kyber.Scalar, p.N)
+	HS := make([]kyber.Point, p.N)
+	for i := 0; i < p.N; i++ {
 		indices[i] = priShares[i].I
 		values[i] = priShares[i].V
 		HS[i] = H
 	}
 
 	// Create NIZK discrete-logarithm equality proofs
-	proofs, _, sX, err := dleq.NewDLEQProofBatch(suite, HS, X, values)
+	proofs, coms, sX, err := dleq.NewDLEQProofBatch(suite, HS, X, values)
 	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
 
-	for i := 0; i < n; i++ {
+	for i := 0; i < p.N; i++ {
 		ps := &share.PubShare{I: indices[i], V: sX[i]}
-		encShares[i] = &PubVerShare{*ps, *proofs[i]}
+		p.Shares[i] = &PubVerShare{*ps, coms[i], *proofs[i]}
 	}
 
-	return encShares, pubPoly, nil
+	return p, nil
 }
 
 // VerifyEncShare checks that the encrypted share sX satisfies
 // log_{H}(sH) == log_{X}(sX) where sH is the public commitment computed by
 // evaluating the public commitment polynomial at the encrypted share's index i.
 func VerifyEncShare(suite Suite, H kyber.Point, X kyber.Point, sH kyber.Point, encShare *PubVerShare) error {
 	if err := encShare.P.Verify(suite, H, X, sH, encShare.S.V); err != nil {
```

Comment:
We should include the verification of the DLEQ proof challenge `c` here. It's perhaps better to do that verification into `VerifyEncShareBatch` directly, since the "encrypted shares" have a common challenge.

Reply:
I moved the common challenge check in `VerifyEncShareBatch`

Reply:
Great point! I agree, it's probably better to add the param to `VerifyEncShare`.